sstables: Fix deletion of partial SSTables

raphaelsc opened the pull request:
If SSTable write fails, it will leave a partial sst which contains
a temporary TOC in addition to other components partially written.
temporary TOC content is written upfront, to allow us from deleting
all partial components using the former content if write fails.

After commit e5fc4b6, partial sst cannot be deleted because it is
incorrectly assuming all files being deleted unconditionally has
TOC, but that's not true for partial files that need to be removed.
The consequence of this is that space of partial files cannot be
reclaimed, making it worse for Scylla to recover from ENOSPC,
which could happen by selecting a set of files for compaction with
higher chance of suceeeding given the free space.
Let's fix this by taking into account temp TOC for partial files.

Fixes #10410.

Signed-off-by: Raphael S. Carvalho <raphaelsc@scylladb.com>

bhalevy commented:
I thought we had a unit test for that...
Can you please add one?

raphaelsc commented:
> I thought we had a unit test for that... Can you please add one?

I can definitely add one. will send v2 soon. thanks

nyh reviewed (commented):
I guess I must be missing something, but I don't understand what this patch is changing, if at all...
As far as I can tell, new_toc_name and sstable_toc_name have the same directory - to calculate new_toc_name we just take sstable_toc_name, remove the TOC_SUFFIX from the end and add TEMPORARY_TOC_SUFFIX at the end instead.
So dirname(new_toc_name) and dirname(sstable_toc_name) should be the same thing, and this patch shouldn't be changing anything...
What am I missing?

bhalevy commented:
@nyh the root cause is that our implementation of `dirname` calls fs::path::canonical rather than weakly_canonical so it requires the file to exist, not only its parent directory.

nyh commented:
> @nyh the root cause is that our implementation of `dirname` calls fs::path::canonical rather than weakly_canonical so it requires the file to exist, not only its parent directory.

Ok, I wish the commit message said that.

Anyway, in that case it looks to me like the simplest solution would be to replace all three lines added in this patch with just one line - right after the end of the if():
```c++
sstring dir = dirname(new_toc_name)
```
The whole point of the code above it was to make sure that the name "new_toc_name" points to the file we need. so dirname(new_toc_name) is exactly what we need, unconditionally. Much simpler, I think.

raphaelsc commented:
> @nyh the root cause is that our implementation of `dirname` calls fs::path::canonical rather than weakly_canonical so it requires the file to exist, not only its parent directory.

I am liking the idea of just switching to weakly_canonical(). "A normal path of the form canonical(x)/y, where x is a path composed of the longest leading sequence of elements in p that exist, and y is a path composed of the remaining trailing non-existent elements of p"

seems exactly what we need.

raphaelsc commented:
v2: moved dirname() to after conditional block; added unit test.

bhalevy commented on sstables/sstables.cc:
```diff
@@ -2678,21 +2678,21 @@ fsync_directory(const io_error_handler& error_handler, sstring fname) {
         // If new_toc_name exists it will be atomically replaced.  See rename(2)
         co_await sstable_io_check(sstable_write_error_handler, rename_file, sstable_toc_name, new_toc_name);
-        co_await fsync_directory(sstable_write_error_handler, dir);
+        co_await fsync_directory(sstable_write_error_handler, new_toc_name);
```
wouldn't that call fsync on the file rather than the directory?
I think this should be:
```
co_await fsync_directory(sstable_write_error_handler, dirname(sstable_toc_name));
```

raphaelsc replied:
fsync_directory() calls dirname() on file param. I added a sentence in the commit msg explaining that. So I think it's correct to call fsync_directory() with new_toc_name, which all path components do exist.

raphaelsc replied:
fsync_directory() is defined like this:
```
future<>
fsync_directory(const io_error_handler& error_handler, sstring fname)
```

then its open dir like this:
`        return open_checked_directory(error_handler, dirname(fname)).then([] (file f) {`

bhalevy replied:
ah, ok, then it will work then. Sorry about the noise

nyh replied:
So it's not just that the old code sometimes passed the wrong thing, it always passed the wrong thing because it passed the parent directory instead of a file in that directory as expected by the fsync_directory function?
By the way, fsync_directory() should be renamed fsync_parent() - it's too confusing :-(

raphaelsc replied:
> So it's not just that the old code sometimes passed the wrong thing, it always passed the wrong thing because it passed the parent directory instead of a file in that directory as expected by the fsync_directory function? 

Hi @nyh,

Prior to this patch, if remove_by_toc_name() were called on a SSTable containing only TMP TOC

an exception would be triggered in this line: https://github.com/scylladb/scylla/blob/bce2933d99d5e5b86fb0cfccb2882f0b1710db8d/sstables/sstables.cc#L2681

because TOC doesn't exist, meaning fsync_directory() wouldn't be called.

If TOC exists, for a sealed SSTable, then fsync_directory() is called on the directory which the SSTable lives in. Which is exactly what we want to guarantee filesystem has properly updated directory metadata.

> By the way, fsync_directory() should be renamed fsync_parent() - it's too confusing :-(

If I pass /foo/bar, where bar is a directory, I expect fsync_directory() to run on /foo/bar, not on /foo.

If I instead pass /foo/bar, where bar is a regular file, it still makes sense that fsync_directory() picks the underlying directory, which is /foo.

Which makes me think fsync_directory() naming is good as-is.

After my patch, fsync_directory() is now called with absolute path to TMP TOC, meaning we still want it to sync the underlying directory, the directory which TMP TOC lives in.

bhalevy replied:
> So it's not just that the old code sometimes passed the wrong thing, it always passed the wrong thing because it passed the parent directory instead of a file in that directory as expected by the fsync_directory function?

Right.

> By the way, fsync_directory() should be renamed fsync_parent() - it's too confusing :-(

I agree with @raphaelsc that it can be fixed to what its name suggests it does (and documenting that) by fstat'ing `fname` to determine if it's a directory or not and fsync'ing its parent directory only if it is not a directory by itself.

But I'm not sure it's worth the extra logic, since the caller knows what it wants to fsync and can pass the actual directory to fsync.  Note that we're still doing the wrong thing at the bottom of this function:
https://github.com/scylladb/scylla/blob/bcd35af7cf93eac9664a044816847521cfddb0ac/sstables/sstables.cc#L2731

nyh replied:
> If TOC exists, for a sealed SSTable, then fsync_directory() is called on the directory which the SSTable lives in. Which is exactly what we want to guarantee filesystem has properly updated directory metadata.

My comment here was about the **old** (before patch) code being even more wrong that we thought it is (unless I misunderstood something): The old code had:
```c++
auto dir = dirname(sstable_toc_name);
co_await fsync_directory(sstable_write_error_handler, dir);
```

If I understand correctly, if sstable_toc_name is "/a/b/c.toc", then "dir" is "/a/b" and fsync_directory(dir) takes dirname(dir) and gets "/a" and then does a fsync on this "/a". Wasn't that completely wrong (regardless of whether we had toc or toc.tmp)?

bhalevy replied:
@nyh I agree with you.  The old implementation was wrong and the current version of this PR is still wrong.
I suggest to fix that as part of this PR by taking the `dirname` call out of `fsync_directory` so its caller will make sure to call it with the right directory to fsync.  There are only 2 call sites: one after the rename of TOC to TemporaryTOC and a final one after we're done removing all components.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/benipeled_test/28/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/157/

bhalevy reviewed (changes requested):
The final call to fsync_directory is still wrong since we pass `dir` to it rather than `new_toc_name`.

I suggest to fix that as part of this PR by taking the dirname call out of fsync_directory so its caller will make sure to call it with the right directory to fsync. There are only 2 call sites: one after the rename of TOC to TemporaryTOC and a final one after we're done removing all components.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/177/

bhalevy commented on sstables/sstables.cc:
```diff
@@ -2659,14 +2659,14 @@ future<> sstable::close_files() {
 }
 
-static inline sstring dirname(const sstring& fname) {
+static inline sstring parent_path(const sstring& fname) {
```
no objection since `parent_path` is a better name for this function, just for the protocol, `dirname` follows the posix convention of dirname/basename.  See `man 3 dirname`

nyh replied:
Right. I think dirname was good too. I think the confusing name wasn't of dirname() but but of fsync_directory() which sounded like it was acting on a directory path - not its parent. But you fix that in the next patch.

bhalevy reviewed (approved):
LGTM, thanks

bhalevy commented:
@nyh please re-review and consider merging

nyh commented on sstables/sstables.cc:
```diff
@@ -2729,7 +2728,7 @@ remove_by_toc_name(sstring sstable_toc_name) {
         }
     });
-    co_await fsync_directory(sstable_write_error_handler, dir);
+    co_await fsync_directory(sstable_write_error_handler, parent_path(new_toc_name));
```
I still I don't understand why in the first case of the if(), you want to run exactly the same fsync_directory() call twice - once at the end of the if(), and once here right after ending the if! What am I missing?

bhalevy replied:
If we crash, we want to make sure the TOC file was renamed to the temporary TOC name, first, before deleting any other components, since this is the commit point.
We want to avoid ending up with the TOC name and missing components.

raphaelsc replied:
It's one fsync after renaming TOC to TMP TOC, and one fsync after removing all components *and* before removing TMP TOC.

raphaelsc replied:
@nyh See commit for a simplified form of this function when there was no TMP TOC: 851248e263256069768e96f32494a8fe999bc203
